src/gui/KM_GUIMenuOptions.pas

```diff
@@ -525,21 +482,36 @@ procedure TKMMenuOptions.KeybindClick(Sender: TObject);
 
 procedure TKMMenuOptions.Keybind_ListKeySave(Key: Word; Shift: TShiftState);
 var
-  aID, I: Integer;
+  ID, I: Integer;
+  fi: TKMFuncInfo;
 begin
   ColumnBox_Options_Keys.HighlightError := False;
-  aID := ColumnBox_Options_Keys.ItemIndex;
-  // Never allow to change secret debug keys.
-  if (aID >= 0) and (aID <= gResKeys.Count - 5) then
+  ID := ColumnBox_Options_Keys.Rows[ColumnBox_Options_Keys.ItemIndex].Tag;
+  // Never allow to set F10 or F11 keys.
+  if (ID >= 0) and (ID <= gResKeys.Count - 1) then
   begin
-    for I := 0 to gResKeys.Count -1 do
-      if (Key = gResKeys.Keys[I]) or (Key in [121, 122]) then
-      begin
-        ColumnBox_Options_Keys.HighlightError := True;
-        gSoundPlayer.Play(sfxn_Error);
-        Exit;
-      end;
-    gResKeys.Keys[aID] := Key;
+    if not gResKeys.AllowKeySet(Key) then
+    begin
+      ColumnBox_Options_Keys.HighlightError := True;
+      gSoundPlayer.Play(sfxn_Error);
+      Exit;
+    end;
+
+    for I := 0 to gResKeys.Count - 1 do
```

Comment:
So far this is what's working of the function area check, if the function is in the kaCommon area and there is a key with the same value then the one you pressed, replace the existing function with key value 0 and set the function you wish to the pressed key value.